### ISSUE TITLE
WL MAXIMIZE do not work on MAC M1 QEMU

### ISSUE THREAD

**Issue by munoz0raul** [Igalia/cog#537]
Hi All, 

I'm using QEMU and COG and on my Intel machine, ruing QEMU ARM64 everything works fine. But on a college machine with the MAC M1 (ARM), running QEMU ARM64, maximize does not work, and when I remove maximize and set the QEMU resolution the cog is launched with an offset. All other Wayland Apps I have work fine in both environments.

Here are some examples:

```
COG_PLATFORM_WL_VIEW_MAXIMIZE="0"
COG_PLATFORM_WL_VIEW_WIDTH="1280"
COG_PLATFORM_WL_VIEW_HEIGHT:="768"
```

Intel x86 (Linux) running QEMU ARM: (Works fine)
![image](https://user-images.githubusercontent.com/7702208/219238837-099cee93-d5fe-497a-a9bc-70b95d0404cd.png)

MAC M1 ARM64 (MAC OS) running QEMU ARM: (Offset)
![image](https://user-images.githubusercontent.com/7702208/219239286-48776ec9-e5a0-4a45-884e-23a40418e864.png)

When trying fullscreen:

`COG_PLATFORM_WL_VIEW_MAXIMIZE="1"
`

Intel x86 (Linux) running QEMU ARM: (Works fine)
![image](https://user-images.githubusercontent.com/7702208/219240135-b74f9b86-2147-466e-86c7-cc798aa774a2.png)

MAC M1 ARM64 (MAC OS) running QEMU ARM: (Error)

```
(cog:8): GLib-GIO-WARNING **: 00:36:50.081: Your application does not implement g_application_activate() and has no handlers connected to the 'activate' signal.  It should do one of these.
Cog-Core-Message: 00:36:50.154: <file:///index.html> Load started.
xdg_wm_base@11: error 4: xdg_surface geometry (1024 x 768) does not match the configured maximized state (1280 x 768)
Cog-Core-Message: 00:36:50.157: <file:///index.html> Loading...
Cog-Core-Message: 00:36:50.162: <file:///index.html> Loaded successfully.
```

Funny that FULLSCREEN works fine for both environments.

I'm wondering if there is something implemented for the FULLSCREEN that was not added to the MAXIMIZE.

I'm using COG upstream built with Yocto.

**Comment by munoz0raul** [comment]
I found the problem. When using QEMU on mac, it doesn't have EGL, so Weston starts without EGL:

`EGL Wayland extension: no`

Instead of EGL, it uses the SHM buffer, which doesn't get the correct resolution when using maximized.

https://github.com/Igalia/cog/blob/3d579f10304e66d94601142f2c005985cf39b069/platform/wayland/cog-platform-wl.c#L1807

Also, it isn't very clear to me if width and height should use the complete screen when maximized or if it should subtract the Weston panel bar.

The solution I found was to read the width and height from the environment variables. So I have a way to manipulate the values depending on the error I get:

`xdg_wm_base@11: error 4: xdg_surface geometry (1024 x 768) does not match the configured maximized state (1280 x 768)`

This is my final solution:

```
static void
on_export_shm_buffer (void* data, struct wpe_fdo_shm_exported_buffer* exported_buffer)
...
...
...
        if (win_data.is_fullscreen) {
            g_debug("LINE:%d \n", __LINE__);
            width = win_data.width;
            height = win_data.height;
        } else {
            g_debug("LINE:%d \n", __LINE__);
-            width = wl_shm_buffer_get_width(exported_shm_buffer);
-            height = wl_shm_buffer_get_height(exported_shm_buffer);
+            configure_surface_geometry(0, 0);
+            width = win_data.width;
+            height = win_data.height;
        }
```

And as enviroment variables:

```
COG_PLATFORM_WL_VIEW_MAXIMIZE: "1"
COG_PLATFORM_WL_VIEW_WIDTH: "1280"
COG_PLATFORM_WL_VIEW_HEIGHT: "768"
```

If someone knows the right way to get the real width and height from Weston, I can try to test and suggest a patch.

**Comment by munoz0raul** [comment]
What is funny is that on some devices I get the suggested height as 736, sometimes 768. Probably because of the top bar:

`xdg_wm_base@11: error 4: xdg_surface geometry (1280 x 768) does not match the configured maximized state (1280 x 736)`

**Comment by alexgcastro** [comment]
Thanks for the report! Does this patch solve the problem for you: https://github.com/Igalia/cog/pull/539

**Comment by lauromoura** [comment]
> Thanks for the report! Does this patch solve the problem for you: #539

Testing on a x86-64 Ubuntu 22.04 (GNOME Wayland), #539 indeed made Cog respect `COG_PLATFORM_WL_VIEW_MAXIMIZE` and related size vars. Without it, it would always start at 1024x768.

**Comment by aperezdc** [comment]
@munoz0raul I am going to tentatively close this as resolved by #539 but if you build a version of Cog with this patch and you still experience the same issue, please reopen this bug report and we'll take another look. Thanks!
